File: Algorithms/p003_Longest_Substring_Without_Repeating_Characters/p003_Longest_Substring_Without_Repeating_Characters_2.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        index = {chr(i):-1 for i in range(256)} # index[c]: index of c in s so far
        max_len = 0
        start = 0
        for i in range(len(s)):
            start = max(start,index[s[i]]+1)
            max_len = max(max_len,i-start+1)
            index[s[i]] = i
        return max_len

File: Algorithms/p003_Longest_Substring_Without_Repeating_Characters/p003_Longest_Substring_Without_Repeating_Characters_2.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> index(256,-1); // index[c]: index of c in s so far
        int max_len = 0;
        for(int i=0,start=0;i<s.size();++i){
            start = max(start,index[s[i]]+1);
            max_len = max(max_len,i-start+1);
            index[s[i]] = i;
        }
        return max_len;
    }
};
